vmware: provide instructions for modifying OVA (#376)
coreos/fedora-coreos-docs: modules/ROOT/pages/provisioning-vmware.adoc

This guide shows how to provision new Fedora CoreOS (FCOS) nodes on the VMware hypervisor.

NOTE: Only https://kb.vmware.com/s/article/1003746[hardware version] 13 or later, vSphere ESXi hosts version 6.5 or later and vCenter host version 6.5 or later are supported. It may be possible to

> xref:modifying-ovf-metadta.adoc[modify the metadata of the OVA] to run in older versions of vSphere or vCenter,

Reviewer: This mentions OVA, but the title of the other document we're adding mentions OVF.. As someone dumb to the vmwhare sphere, what's the difference?

Reviewer: Ahh I see a paragraph in the new doc about it "shipped as an OVA, which is simply a tarball that contains the files disk.vmdk and coreos.ovf"

Reviewer: in that case maybe we can say "modify the OVF metadata of the OVA" here.

govc device.serial.add -vm "${VM_NAME}"
govc device.serial.connect -vm "${VM_NAME}" "[datastore] ${VM_NAME}/console.log"
----

> == Modifying OVA metadata

Reviewer 2: Let's try to distinguish OVA and OVF throughout this section.  The metadata is OVF and the single-file archive is OVA.

NOTE: While we provide these instructions for modifying the OVA, we cannot
guarantee that any modifications to the OVA metadata will result in a usable
guest VM.

Fedora CoreOS ships the VMware OVA with certain metadata defaults in the OVA file.
The hardware version and default firmware are of particular interest, as they
can affect how the guest VM operates in the hypervisor. These defaults may

> prevent the use of FCOS as a guest VM in older versions of VMware products and

Reviewer 2: Should we explicitly state a policy that we will support non-EOL versions of VMware products?

Author: I think that is fair; though we don't have any regular testing on those platforms.

Author: Is this something we should discuss more broadly before that kind of statement?

Reviewer 2: > I think that is fair; though we don't have any regular testing on those platforms.

I meant in the limited sense that we won't bump the HW version to exclude non-EOL products.  While we're telling users how to change the defaults, it may be helpful to mention when they won't need to.

> Is this something we should discuss more broadly before making that kind of statement?

Probably.

it may be necessary to modify the metadata in order to use FCOS as a guest VM.

The VMware OVA is simply a tarball that contains the files `disk.vmdk` and
`coreos.ovf`. In order to edit the metadata used by FCOS as a guest VM, you

> should untar the OVA artifact, edit the OVF file, then create a new OVA file.

Reviewer 2: In principle the OVA should also include a `.mf` file, which is a list of SHA digests for the other files.  We're not doing that because when I tried, I got mysterious hash mismatches I couldn't track down.  I don't think it's especially important, but wanted to note that these instructions assume the user doesn't need to update or remove an `.mf` file after changing the `.ovf`.

In the example below, the hardware version in the OVF is changed from the

> default value of `17` to a newer version of `13`. (Note: the defaults in the

Reviewer 2: Did 13 and 17 get swapped here?

Author: Oops, good catch.

Author: Actually, I am just going to update the wording to align with the example of changing from a newer default value to an older value.

govc device.serial.add -vm "${VM_NAME}"
govc device.serial.connect -vm "${VM_NAME}" "[datastore] ${VM_NAME}/console.log"
----

== Modifying OVF metadata

NOTE: While we provide these instructions for modifying the OVF metadata, we cannot
guarantee that any modifications to the OVF metadata will result in a usable
guest VM.

> Fedora CoreOS ships the VMware OVA with metadata defaults defined in the OVF
> file contained within the OVA file. The hardware version and default firmware
> are of particular interest, as they can affect how the guest VM operates in the
> hypervisor. These defaults may prevent the use of FCOS as a guest VM in older
> versions of VMware products and it may be necessary to modify the metadata in
> order to use FCOS as a guest VM.
> 
> NOTE: The Fedora CoreOS project will not make any changes to the hardware
> version in the OVF metadata of the VMware OVA that would exclude non-EOL
> platforms from being supported.

Reviewer 2: Maybe something like this?

```suggestion
Fedora CoreOS is intended to run on
https://lifecycle.vmware.com/[generally supported] releases of VMware ESXi,
VMware Workstation, and VMware Fusion. Accordingly, the Fedora CoreOS VMware
OVA image specifies a
https://kb.vmware.com/s/article/1003746[virtual hardware version]
that may not be compatible with older, unsupported VMware products.
However, you can modify the image's OVF metadata to specify an older
virtual hardware version.
```

The VMware OVA is simply a tarball that contains the files `disk.vmdk` and
`coreos.ovf`. In order to edit the metadata used by FCOS as a guest VM, you
should untar the OVA artifact, edit the OVF file, then create a new OVA file.

> In the example below, the hardware version in the OVF is changed from the

Reviewer 2: ```suggestion
The example commands below change the OVF hardware version from the
```

default value of `17` to an older version of `13`. (Note: the defaults in the
OVF are subject to change.)

> [source, bash]
> ----
> tar -xvf fedora-coreos-35.20220213.3.0-vmware.x86_64.ova
> sed -i 's|vmx-17|vmx-13|' coreos.ovf
> tar -H posix -cvf fedora-coreos-35.20220213.3.0-vmware-vmx-13.x86_64.ova coreos.ovf disk.vmdk
> ----

Reviewer 2: Let's avoid breaking user scripts when we change the default, and avoid hardcoding version numbers.

```suggestion
[source,bash,subs="attributes"]
----
tar -xvf fedora-coreos-{stable-version}-vmware.x86_64.ova
sed -iE 's/vmx-[0-9]*/vmx-13/' coreos.ovf
tar -H posix -cvf fedora-coreos-{stable-version}-vmware-vmx-13.x86_64.ova coreos.ovf disk.vmdk
----
```
